packages/odyssey-react/src/components/TextInput/TextInput.tsx

```diff
@@ -68,14 +69,40 @@ interface CommonProps
   placeholder?: string;
 
   /**
-   * An optional string prefix displayed before the input
+   * An optional prefix button displayed before the input
+   * Only one of PrefixButton, PrefixIcon, PrefixText can be set
    */
-  prefix?: string;
+  PrefixButton?: ReactNode;
 
   /**
-   * An optional string suffix displayed after the input
+   * An optional prefix icon displayed before the input
+   * Only one of PrefixButton, PrefixIcon, PrefixText can be set
    */
-  suffix?: string;
+  PrefixIcon?: ReactNode;
+
+  /**
+   * An optional prefix text displayed before the input
+   * Only one of PrefixButton, PrefixIcon, PrefixText can be set
+   */
+  PrefixText?: string;
+
+  /**
+   * An optional suffix button displayed before the input
+   * Only one of SuffixButton, SuffixIcon, SuffixText can be set
+   */
+  SuffixButton?: ReactNode;
+
+  /**
+   * An optional suffix icon displayed before the input
+   * Only one of SuffixButton, SuffixIcon, SuffixText can be set
+   */
+  SuffixIcon?: ReactNode;
+
+  /**
+   * An optional suffix text displayed before the input
+   * Only one of SuffixButton, SuffixIcon, SuffixText can be set
+   */
+  SuffixText?: string;
```

Comment:
I'm not sure if you saw the thread from last week but we can do a run time check against children to determine the type of children!

Reply:
Do we need to know at run time?
The way the logic is right now, it just allows one prefix/suffix and chooses in order of Text, Icon and then Button

Reply:
yea this is the more granular API we talked about building if we couldn't get the unified polymorphic API to work. It turns out we can use one of two approaches: introspect on the component constructor, or use sentinel props. While not ideal, I'm not too worried about the API thrash if we want to merge this then refactor in the future that is fine too.